### PR TITLE
feat: support <driver/i2c_master.h> in i2cdev

### PR DESCRIPTION
**Pull request by trombik**
This is a premature PoC, calling for feedback. I did not test it with a real hardware.  There should be style violations and other run-time errors. Never ported any dependent I2C components. Successfully compiled `examples/i2cdev/default` with and without `CONFIG_I2CDEV_USING_LEGACY_I2C`. I just followed instructions for new `i2c_master`, kept the original logic as much as possible, and made it compatible with older implementation.

> I certainly agree that it’s time to port i2clib to a new driver (and at the same time maintain compatibility with older versions of ESP-IDF and ESP8266 RTOS SDK).
> 
> I invite the discussion participants (and everyone else) to initiate the process by creating a PR and starting work in it.

_Originally posted by @UncleRus in https://github.com/UncleRus/esp-idf-lib/discussions/612#discussioncomment-8579750_

Questions:

* Is the effort going in the right direction?
* Any fundamental design mistakes?

### CODE REVIEW

File: components/i2cdev/i2cdev.h [hunk 1]
```diff
@@ -68,12 +73,23 @@ extern "C" {
...
                                   ticks for ESP-IDF, CPU ticks for ESP8266.
-                                  When this value is 0, I2CDEV_MAX_STRETCH_TIME will be used */
+                                  When this value is 0, I2CDEV_MAX_STRETCH_TIME will be used
+                                  (i2c.h only) */
+#else
+    i2c_master_bus_handle_t bus; //!< I2C master bus handle (i2c_master.h only)
+    i2c_master_dev_handle_t device; //!< I2C master bus device handle
+                                    //(i2c_master.h only)
+    i2c_device_config_t device_config; //!< I2C device configuration (i2c_master.h only)
+    i2c_master_bus_config_t master_bus_config; /*! < I2C master bus specific
```
Comment (camielverdult) [inline components/i2cdev/i2cdev.h:87]: It understand that the i2cdev component could stay in charge of the i2c bus, such as it was when using the previous I2C driver. Espressif added the option for queuing and thread safety, allowing multiple components/libraries to share the bus. Taking ownership of a bus to (again) be used by only this component creates the same conflicts as with the previous version: needing to either use a different bus for a device that is not supported by esp-idf-lib or (now that a bus can be shared) "digging" into this (i2cdev) component for another component to be able to share the bus. In my (very junior) opinion, would it not provide more flexibility to let the develop instantiate the bus (as you already now in the main.c file) and pass it as a pointer, for i2cdev to access (too)?

File: components/i2cdev/i2cdev_master.c [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        {
+            res = i2c_master_receive(dev->device, (void *)out_data, out_size, CONFIG_I2CDEV_TIMEOUT);
+        }
+        if (res != ESP_OK)
+        {
+            ESP_LOGE(TAG, "Could not read from device [0x%02x at %d]: %d (%s)", dev->addr, dev->port, res, esp_err_to_name(res));
+        }
+    }
+
+    SEMAPHORE_GIVE(dev->port);
```
Comment (camielverdult) [inline components/i2cdev/i2cdev_master.c:148]: Why would we lock the bus when using the updated driver? It seems that this feature is already present in the bus structure ([link](https://github.com/espressif/esp-idf/blob/master/components/esp_driver_i2c/i2c_private.h#L131))

File: examples/i2cdev/default/main/main.c [hunk 3]
```diff
@@ -6,9 +6,20 @@
...
-#if HELPER_TARGET_IS_ESP32
+#else
+    dev.master_bus_config.sda_io_num = CONFIG_EXAMPLE_I2C_MASTER_SDA;
+    dev.master_bus_config.scl_io_num = CONFIG_EXAMPLE_I2C_MASTER_SCL;
+    dev.master_bus_config.i2c_port = -1;
+    dev.master_bus_config.clk_source = I2C_CLK_SRC_DEFAULT;
+    dev.master_bus_config.intr_priority = 0;
+    dev.master_bus_config.glitch_ignore_cnt = 7;
+    dev.master_bus_config.trans_queue_depth = 0;
+    dev.master_bus_config.flags.enable_internal_pullup = false;
```
Comment (camielverdult) [inline examples/i2cdev/default/main/main.c:20]: Using the i2c_dev component in this way locks the user/programmer into using this component for all I2C actions. I think this doesn't fit with the features' of the updated I2C driver (queuing). Wouldn't it be much more flexible for this component to store a pointer instead of possessing the bus (again)?

**Review by camielverdult — COMMENTED** [review]
Hello,

Heads up: I am not from esp-idf-lib (and also not a senior developer). This new version looks great and would be very useful. My comments are focused on the integration of the i2c_master_bus into the existing i2c_dev_t structure. This updated I2C driver from espressif allows for multiple components to share the bus, without the need for an extra layer orchestrating the access between threads. The proposed code in this commit takes away the opportunity to use the new features' of the updated I2C driver, forcing developers to use a second I2C bus (again) if they prefer integration between components without needing to dive into the code of esp-idf-lib's i2cdev component. 

tl;dr: I don't think taking ownership of the bus (again) is a good idea given the queuing approach from espressif.

But, I could be completely wrong. You probably have researched this a little more than I have, so I am very curious about your thoughts.

**Comment by camielverdult** [comment]
Also, a lot of components (using i2cdev) are still using an initializer function using the following:

```c
esp_err_t xxx_init_desc(aht_t *dev, uint8_t addr, i2c_port_t port, gpio_num_t sda_gpio, gpio_num_t scl_gpio)
{
    CHECK_ARG(dev);

    ...

    // Using legacy 
    dev->i2c_dev.addr = addr;
    dev->i2c_dev.cfg.sda_io_num = sda_gpio;
    dev->i2c_dev.cfg.scl_io_num = scl_gpio;
#if HELPER_TARGET_IS_ESP32
    dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
```

<details closed>
<summary>All of the files</summary>
<br>
21 files

build/_deps/espidflib-src/components/aht/aht.c:
  90      dev->i2c_dev.port = port;
  91:     dev->i2c_dev.addr = addr;
  92:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  93:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  94: #if HELPER_TARGET_IS_ESP32
  95:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  96  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/aht/aht.c:
  90      dev->i2c_dev.port = port;
  91:     dev->i2c_dev.addr = addr;
  92:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  93:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  94: #if HELPER_TARGET_IS_ESP32
  95:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  96  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/bh1900nux/bh1900nux.c:
  141      dev->i2c_dev.port = port;
  142:     dev->i2c_dev.addr = addr;
  143:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  144:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  145: #if HELPER_TARGET_IS_ESP32
  146:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  147  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/bme680/bme680.c:
  566      dev->i2c_dev.port = port;
  567:     dev->i2c_dev.addr = addr;
  568:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  569:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  570: #if HELPER_TARGET_IS_ESP32
  571:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  572  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/bmp280/bmp280.c:
  161      dev->i2c_dev.port = port;
  162:     dev->i2c_dev.addr = addr;
  163:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  164:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  165: #if HELPER_TARGET_IS_ESP32
  166:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  167  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/ccs811/ccs811.c:
  268      dev->i2c_dev.port = port;
  269:     dev->i2c_dev.addr = addr;
  270:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  271:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  272: #if HELPER_TARGET_IS_ESP32
  273:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  274  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/hdc1000/hdc1000.c:
  137      dev->i2c_dev.port = port;
  138:     dev->i2c_dev.addr = addr;
  139:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  140:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  141: #if HELPER_TARGET_IS_ESP32
  142:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  143  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/ina219/ina219.c:
  125      dev->i2c_dev.port = port;
  126:     dev->i2c_dev.addr = addr;
  127:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  128:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  129: #if HELPER_TARGET_IS_ESP32
  130:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  131  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/ina260/ina260.c:
  124      dev->i2c_dev.port = port;
  125:     dev->i2c_dev.addr = addr;
  126:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  127:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  128: #if HELPER_TARGET_IS_ESP32
  129:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  130  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/ina3221/ina3221.c:
  105      dev->i2c_dev.port = port;
  106:     dev->i2c_dev.addr = addr;
  107:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  108:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  109: #if HELPER_TARGET_IS_ESP32
  110:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  111  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/l3gx/l3gx.c:
  141      dev->i2c_dev.port = port;
  142:     dev->i2c_dev.addr = addr;
  143:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  144:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  145: #if HELPER_TARGET_IS_ESP32
  146:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  147  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/mcp342x/mcp342x.c:
  110      dev->i2c_dev.port = port;
  111:     dev->i2c_dev.addr = addr;
  112:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  113:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  114: #if HELPER_TARGET_IS_ESP32
  115:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  116  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/mcp960x/mcp960x.c:
  181      dev->i2c_dev.port = port;
  182:     dev->i2c_dev.addr = addr;
  183:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  184:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  185: #if HELPER_TARGET_IS_ESP32
  186:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  187  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/mpu6050/mpu6050.c:
  206      dev->i2c_dev.port = port;
  207:     dev->i2c_dev.addr = addr;
  208:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  209:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  210: #if HELPER_TARGET_IS_ESP32
  211:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  212  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/ms5611/ms5611.c:
  167      dev->i2c_dev.port = port;
  168:     dev->i2c_dev.addr = addr;
  169:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  170:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  171: #if HELPER_TARGET_IS_ESP32
  172:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  173  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/qmc5883l/qmc5883l.c:
  104      dev->i2c_dev.port = port;
  105:     dev->i2c_dev.addr = addr;
  106:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  107:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  108: #if HELPER_TARGET_IS_ESP32
  109:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  110  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/qmp6988/qmp6988.c:
  286      dev->i2c_dev.port = port;
  287:     dev->i2c_dev.addr = addr;
  288:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  289:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  290: #if HELPER_TARGET_IS_ESP32
  291:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  292  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/sht3x/sht3x.c:
  210      dev->i2c_dev.port = port;
  211:     dev->i2c_dev.addr = addr;
  212:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  213:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  214: #if HELPER_TARGET_IS_ESP32
  215:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  216  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/sts3x/sts3x.c:
  205      dev->i2c_dev.port = port;
  206:     dev->i2c_dev.addr = addr;
  207:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  208:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  209: #if HELPER_TARGET_IS_ESP32
  210:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  211  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/tsl2561/tsl2561.c:
  214      dev->i2c_dev.port = port;
  215:     dev->i2c_dev.addr = addr;
  216:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  217:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  218: #if HELPER_TARGET_IS_ESP32
  219:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  220  #endif

esp-idf-ssd1306/components/ssd1306/espidflib-src/components/tsys01/tsys01.c:
  104      dev->i2c_dev.port = port;
  105:     dev->i2c_dev.addr = addr;
  106:     dev->i2c_dev.cfg.sda_io_num = sda_gpio;
  107:     dev->i2c_dev.cfg.scl_io_num = scl_gpio;
  108: #if HELPER_TARGET_IS_ESP32
  109:     dev->i2c_dev.cfg.master.clk_speed = I2C_FREQ_HZ;
  110  #endif

</details>

**Comment by camielverdult** [comment]
I've started working on an example for splitting the functionality: https://github.com/camielverdult/esp-idf-lib/commit/49722ccb0ca14de94a7549da79076f5237e1d6f4

File: components/i2cdev/i2cdev_master.c [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        memcpy(write_buffer, out_data, out_size);
+    }
+    res = i2c_master_transmit(dev->device, write_buffer, out_reg_size + out_size, CONFIG_I2CDEV_TIMEOUT);
+    if (res != ESP_OK)
+    {
+        ESP_LOGE(TAG, "Could not write to device [0x%02x at %d]: %d (%s)", dev->addr, dev->port, res, esp_err_to_name(res));
+        goto fail;
+    }
+fail:
+    free(write_buffer);
```
Comment (camielverdult) [inline components/i2cdev/i2cdev_master.c:194]: Are you only freeing on fail? Doesn't this leak memory when the transaction succeeds?

File: components/i2cdev/i2cdev_master.c [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    if (!dev || !out_data || !out_size) return ESP_ERR_INVALID_ARG;
+    uint8_t *write_buffer = NULL;
+    size_t write_buffer_size = 0;
+    esp_err_t res = ESP_FAIL;
+    bool with_register_address = out_reg && out_reg_size;
+
+    SEMAPHORE_TAKE(dev->port);
+
+    write_buffer_size = with_register_address ? out_reg_size + out_size : out_size;
+    write_buffer = malloc(write_buffer_size);
```
Comment (camielverdult) [inline components/i2cdev/i2cdev_master.c:163]: I am guessing you're using a heap allocated buffer to prevent a task's (limited) stack from being used instead. Is it not possible to use the stack?

File: components/i2cdev/i2cdev_master.c [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+    SEMAPHORE_TAKE(dev->port);
+
+    esp_err_t res = i2c_setup_port(dev);
+    if (res == ESP_OK)
+    {
+        res = i2c_master_probe(dev->bus, dev->device_config.device_address, CONFIG_I2CDEV_TIMEOUT);
+    }
+
+    SEMAPHORE_GIVE(dev->port);
```
Comment (camielverdult) [inline components/i2cdev/i2cdev_master.c:119]: This is what I meant with the repetition, wouldn't the code be much clearly arranged to only interchange the calls to the old and new i2c drivers (using preprocessor statements)?

**Comment by camielverdult** [comment]
A much more radical idea: What about not implementing the new i2c_master into i2cdev? Instead, device structures might get away with storing `i2c_master_dev_handle_t` instead of esp-idf-lib's `i2c_dev_t`. A downside is that the user would need to setup the master bus themselves, but it would also give them the control over the bus. I think this is much nicer, especially since the current implementation of i2c_dev possesses an entire bus for a single device (which is the opposite of what I2C is so good at).

**Comment by trombik** [comment]
> tl;dr: I don't think taking ownership of the bus (again) is a good idea given the queuing approach from espressif.
> 
> But, I could be completely wrong. You probably have researched this a little more than I have, so I am very curious about your thoughts.

Your comment is truly valid. The way I implemented it does not utilize the new feature you have mentioned. I was not able to think of a better way without loosing backward compatibility. A very hack.

I will reply to your review comments, but here is a general idea: the primary focus is to keep the same interfaces. I don't understand every parts in the original code. The secondary focus is to start discussions, just like you did and which I appreciate.

Keeping the i2cdev.h as is is another idea. What the original author said is "keeping backward compatibility". depending o the definition, it might have a gray zone. Only the original author can answer how.

File: examples/i2cdev/default/main/main.c (hunk 3 above)
Comment (trombik) [reply examples/i2cdev/default/main/main.c:20]: It would be. What I came up is: create an abstraction header, "i2c_bus.h" for writing less code in most of simple cases, and pass a pointer (and whatever it needs to function) to "i2c_master.h". Users are free to choose either "i2c_bus.h" or their own.

File: components/i2cdev/i2cdev.c [hunk 7]
```diff
@@ -173,160 +114,19 @@ esp_err_t i2c_dev_give_mutex(i2c_dev_t *dev)
...
-
-        res = i2c_master_cmd_begin(dev->port, cmd, pdMS_TO_TICKS(CONFIG_I2CDEV_TIMEOUT));
-
-        i2c_cmd_link_delete(cmd);
-    }
-
-    SEMAPHORE_GIVE(dev->port);
-
-    return res;
+    return _i2c_dev_probe(dev, operation_type);
```
[the remark this answers is not in the document]
Comment (trombik) [reply components/i2cdev/i2cdev.c:119]: Completely right.

File: components/i2cdev/i2cdev_master.c (hunk 5 above)
Comment (trombik) [reply components/i2cdev/i2cdev_master.c:163]: As the part is obtained from the old code, I cannot comment on this.

File: components/i2cdev/i2cdev_master.c (hunk 4 above)
Comment (trombik) [reply components/i2cdev/i2cdev_master.c:194]: IIRC, there is no other path to return after `malloc()` even if succeeded or failed.

File: components/i2cdev/i2cdev.h [hunk 8]
```diff
@@ -68,12 +73,23 @@ extern "C" {
...
+                                                    configurations (i2c_master.h
+                                                    only) */
+#endif
+    uint8_t addr;            //!< Unshifted address
+    SemaphoreHandle_t mutex; //!< Device mutex
```
[the remark this answers is not in the document]
Comment (trombik) [reply components/i2cdev/i2cdev.h:92]: As the part is obtained from the old code, I cannot comment on this.

File: components/i2cdev/i2cdev_master.c (hunk 6 above)
Comment (trombik) [reply components/i2cdev/i2cdev_master.c:119]: I am inclined to avoid macro because of readability when the change is not trivial. The reason of separating old and new is that, you can modify both as long as the interface (and other expected behaviors) are kept. Others might not agree, though.

**Comment by trombik** [comment]
> A much more radical idea: What about not implementing the new i2c_master into i2cdev? Instead, device structures might get away with storing `i2c_master_dev_handle_t` instead of esp-idf-lib's `i2c_dev_t`. 

I agree.

**Comment by camielverdult** [comment]
> Keeping the i2cdev.h as is is another idea. What the original author said is "keeping backward compatibility". depending o the definition, it might have a gray zone. Only the original author can answer how.

The author did post this about i2c changes: https://github.com/UncleRus/esp-idf-lib/issues/626#issuecomment-2114567234

> I agree.

It might break some user's applications, but might be a better approach on the long run to let a device (on the I2C bus, INA260 for example) store it's i2c_dev config and handle. Those should be enough to use the device. 

I also tried if the device can access the bus object:

```c
// ina260.h

...

#include "driver/i2c_types.h"
#include "driver/i2c_master.h"

...

/**
 * Device descriptor
 */
typedef struct
{
    i2c_device_config_t i2c_dev_config;
    i2c_master_dev_handle_t i2c_dev_handle;
    uint16_t config;   //!< Current config
    uint16_t mfr_id;   //!< Manufacturer ID
    uint16_t die_id;   //!< Die ID
} ina260_t;


// ina260.c

// Using initialized ina26_t dev
ESP_LOGD(TAG, "[0x%02x@%u] Resetting...", dev->i2c_dev_config.device_address, dev->i2c_dev_handle->master_bus->base.port_num);
```

This threw an error during building:
```
error: invalid use of undefined type 'struct i2c_master_dev_t'
```

Not sure why this happens.

**Comment by camielverdult** [comment]
Just out of curiosity, I tried if the device can access the bus object:

```c
// ina260.h

...

#include "driver/i2c_types.h"
#include "driver/i2c_master.h"

...

/**
 * Device descriptor
 */
typedef struct
{
    i2c_device_config_t i2c_dev_config;
    i2c_master_dev_handle_t i2c_dev_handle;
    uint16_t config;   //!< Current config
    uint16_t mfr_id;   //!< Manufacturer ID
    uint16_t die_id;   //!< Die ID
} ina260_t;


// ina260.c

// Using initialized ina26_t dev
ESP_LOGD(TAG, "[0x%02x@%u] Resetting...", dev->i2c_dev_config.device_address, dev->i2c_dev_handle->master_bus->base.port_num);
```

This threw an error during building:
```
error: invalid use of undefined type 'struct i2c_master_dev_t'
```

Not sure why this happens.

File: components/i2cdev/i2cdev_master.c (hunk 5 above)
Comment (camielverdult) [reply components/i2cdev/i2cdev_master.c:163]: Hmmm, then it would depend on the stack size of the task, set up by the user. Most of the tasks have it set to `configMINIMAL_STACK_SIZE * 8 or 3`, not sure if that would be enough for a temp buffer.

File: components/i2cdev/i2cdev.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (camielverdult) [reply components/i2cdev/i2cdev.h:92]: Oh alright, it may be that they use the same logic for controlling the mutex. We'll have to ask the author

File: components/i2cdev/i2cdev_master.c (hunk 6 above)
Comment (camielverdult) [reply components/i2cdev/i2cdev_master.c:119]: That does make sense, but it may introduce two versions of the "same" logic. My opinion is that it may be less readable, but safer to make changes to (because you won't have to make a change twice, removing the possibility of making a mistake)

File: components/i2cdev/i2cdev_master.c (hunk 4 above)
Comment (camielverdult) [reply components/i2cdev/i2cdev_master.c:194]: Right, forgot that, oops

File: examples/i2cdev/default/main/main.c (hunk 3 above)
Comment (camielverdult) [reply examples/i2cdev/default/main/main.c:20]: Sure, that would be nice

**Comment by trombik** [comment]
> Keeping the i2cdev.h as is is another idea.

The biggest issue is ESP8266 support. As ESP8266 RTOS SDK has been in maintenance mode (minimal fixes only), it will not get updated with new `i2c_master.h`. That means we need to maintain two versions of each i2c driver.

**Comment by camielverdult** [comment]
> > Keeping the i2cdev.h as is is another idea.
> 
> The biggest issue is ESP8266 support. As ESP8266 RTOS SDK has been in maintenance mode (minimal fixes only), it will not get updated with new `i2c_master.h`. That means we need to maintain two versions of each i2c driver.

ESP-IDF doesn't support the ESP8266 right? It will be very difficult to have try to support (IMO) deprecated hardware

**Comment by trombik** [comment]
No, it doesn't. However, ESP8266 RTOS SDK has `i2c.h` and the same interface. That's why we support ESP8266.

**Comment by camielverdult** [comment]
> No, it doesn't. However, ESP8266 RTOS SDK has `i2c.h` and the same interface. That's why we support ESP8266.

Alright, well then it might be much harder to maintain this. I've left my comments, good luck!

**Comment by trombik** [comment]
A maintainer of ESP8266 RTOS SDK ported `i2c_master.h` to ESP8266 (see https://github.com/espressif/esp-idf/issues/14678). I am going to integrate it with this repository soon so that you can test and suggest improvements.
